### ISSUE TITLE
False positives in the `custom-error-definition` rule

### ISSUE THREAD

**Issue by sindresorhus** [sindresorhus/eslint-plugin-unicorn#90]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%2440%20Funded-%2300A156.svg" />](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/90)

With the following error definition:

```js
class AbortError extends Error {
	constructor(message) {
		super();

		if (message instanceof Error) {
			this.originalError = message;
			message = message.message;
		}

		this.name = 'AbortError';
		this.message = message;
	}
}
```

*From https://github.com/sindresorhus/p-retry/blob/f4830d72f85ad0ee122eec9f62c271974e3ccc05/index.js#L5-L20*

I get this:

```
  index.js:6:3
  ✖   6:3  Pass the error message to super().                                  unicorn/custom-error-definition
  ✖  14:3  Pass the error message to super() instead of setting this.message.  unicorn/custom-error-definition
```

The first problem is that the `Pass the error message to super().` failure should not be outputted when the failure below is printed. No point in having two errors for the same thing.

Second, I don't think it should fail in this case at all, as I'm required to pass it to `this.message` as the `message` is assigned after `super()`. Not sure what heuristic to use for this though.

@SamVerschueren Thoughts?

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $40.00)

- [<img src='https://avatars3.githubusercontent.com/u/44827199?v=4' alt='issuehunt' width=24 height=24> issuehunt](https://issuehunt.io/u/issuehunt) ($40.00)

### Submitted pull Requests
- [#669 `custom-error-definition`: Fix duplicated reports on `this.message` assign](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/pull/669)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

**Comment by SamVerschueren** [comment]
If `message` is being altered inside the constructor like in this case in the if statement, it shouldn't fail. Not sure if it can be done and how though. But I can always check.

**Comment by florianb** [comment]
It is probably possible to prove if the message was "used" by checking if anything was assigned or a member-function or -attribute was called between `super()` and the usage of `this.message`.

Give me a hint if you need a helping hand @SamVerschueren.

**Comment by sindresorhus** [comment]
I was thinking we just check whether the variable assigned to `this.message` got something assigned to itself earlier in the scope. I don't think we need to make it more complicated than that.

**Comment by florianb** [comment]
You're right - i somehow thought message could be more than a string.

**Comment by sholladay** [comment]
I don't like having to write out the constructor and super call just to get a custom error name.

I should be able to simply do:

```js
class SecurityError extends Error {}
SecurityError.prototype.name = 'SecurityError';
```

**Comment by sindresorhus** [comment]
@sholladay Yes, that's possible, but having *one* way to write all Error subclasses is beneficial. Makes it easier to add additional things later and consistency.

**Comment by IssueHuntBot** [comment]
@issuehunt has funded $40.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/55832243/issues/90) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/55832243/issues/90).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/r/new) on IssueHunt to raise funds.

**Comment by futpib** [comment]
@sindresorhus Isn't this a better definition for your error that passes the rule?

```js
class AbortError extends Error {
	constructor(message) {
		let originalError;

		if (message instanceof Error) {
			originalError = message;
			message = originalError.message;
		}

		super(message);

		this.name = 'AbortError';
	}
}
```

I thinks the rule is right here.

**Comment by sindresorhus** [comment]
@futpib I agree.

---

We still need to do this though:

> The first problem is that the Pass the error message to super(). failure should not be outputted when the failure below is printed. No point in having two errors for the same thing.

**Comment by fisker** [comment]
I've fixed duplicated reports on that, anything else need to be done to close this?

**Comment by sindresorhus** [comment]
> but anything else need to be done to close this?

I don't think so.
